pkg/keys/filesystem.go

```diff
@@ -0,0 +1,410 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keys
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+var _ EncryptionKeyManager = (*Filesystem)(nil)
+var _ KeyManager = (*Filesystem)(nil)
+var _ SigningKeyManager = (*Filesystem)(nil)
+
+// Filesystem is a key manager that uses the filesystem to store and retrieve
+// keys. It should only be used for local development and testing.
+type Filesystem struct {
+	root string
+	mu   sync.RWMutex
+}
+
+// NewFilesystem creates a new, local, in memory KeyManager.
+func NewFilesystem(ctx context.Context, root string) (*Filesystem, error) {
+	if root != "" {
+		if err := os.MkdirAll(root, 0700); err != nil {
+			return nil, err
+		}
+	}
+
+	return &Filesystem{
+		root: root,
+	}, nil
+}
+
+// NewSigner creates a new signer from the given key. If the key does not exist,
+// it returns an error. If the key is not a signing key, it returns an error.
+func (k *Filesystem) NewSigner(ctx context.Context, keyID string) (crypto.Signer, error) {
+	k.mu.RLock()
+	defer k.mu.RUnlock()
+
+	pth := filepath.Join(k.root, keyID)
+	b, err := ioutil.ReadFile(pth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read signing key: %w", err)
+	}
+
+	pk, err := x509.ParseECPrivateKey(b)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse signing key: %w", err)
+	}
+
+	return pk, nil
+}
+
+// Encrypt encrypts the given plaintext and aad with the key. If the key does
+// not exist, it returns an error.
+func (k *Filesystem) Encrypt(ctx context.Context, keyID string, plaintext []byte, aad []byte) ([]byte, error) {
+	k.mu.RLock()
+	defer k.mu.RUnlock()
+
+	// Find the most recent DEK - that's what we'll use for encryption
+	pth := filepath.Join(k.root, keyID)
+	infos, err := ioutil.ReadDir(pth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list keys: %w", err)
+	}
+	if len(infos) < 1 {
+		return nil, fmt.Errorf("there are no key versions")
+	}
+	var latest os.FileInfo
+	for _, info := range infos {
+		if info.Name() == "metadata" {
+			continue
+		}
+		if latest == nil {
+			latest = info
+			continue
+		}
+		if info.Name() > latest.Name() {
+			latest = info
+		}
+	}
+	if latest == nil {
+		return nil, fmt.Errorf("key %s does not exist", keyID)
```

Comment:
nit: change %s to %q in case keyid has spaces in it for log readability.

Reply:
Done